### PR TITLE
Generate: check that `attention_mask` is 2D

### LINKED ISSUE #33549: 4D attention masks not compatible with generate
### System Info

Latest transformers

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Call generate with a 4d attention mask on a llama model. Currently prepare_inputs_for_generation assumes a 2d mask (though does not check the shape, leading to a hard-to-understand error).

```
    from transformers import LlamaForCausalLM, LlamaConfig
    config = LlamaConfig(hidden_dim=64, num_attention_heads=4, num_hidden_layers=1, vocab_size=20, intermediate_size=128)
    model = LlamaForCausalLM(config)

    model.eval()
    input_ids = torch.ones(1, 10).long()
    attention_mask = torch.zeros((
        1, 1, 10, 10
    ))
    # attention_mask = torch.ones(1, 10)
    model.generate(
        input_ids=input_ids,
        max_new_tokens=2,
        do_sample=True,
        num_return_sequences=2,
        attention_mask=attention_mask,
    )
```

### Expected behavior

Generate should probably raise an informative error when passing a 4d attention mask, telling the user they should implement the required logic in e.g. _update_model_kwargs_for_generate or prepare_inputs_for_generation.

currently the errors are hard to understand, and the lack of support for this is unexpected.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes #33549 

In `generate`, `attention_mask` must be in its 2D format. This PR adds a check for that.

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -1861,6 +1861,10 @@ def generate(
             model_kwargs["attention_mask"] = self._prepare_attention_mask_for_generation(
                 inputs_tensor, generation_config._pad_token_tensor, generation_config._eos_token_tensor
             )
+        elif kwargs_has_attention_mask:
+            # TODO (joao): generalize this check with other types of inputs
+            if model_input_name == "input_ids" and len(model_kwargs["attention_mask"].shape) > 2:
```
Comment (gante) [inline src/transformers/generation/utils.py:1866]: I'm not sure `len(model_kwargs["attention_mask"].shape) > 2` generalizes (e.g. some audio models have 3d inputs), so I'm fixing the most common use case now and adding a TODO in case this issue pops up again for other input formats

**Review by LysandreJik — APPROVED** [review]
Makes sense!
